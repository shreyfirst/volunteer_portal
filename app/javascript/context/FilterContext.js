import React, { useState, useCallback } from 'react'

import * as R from 'ramda'

const initialState = {
  houseFilter: { value: undefined },
  eventFilter: { value: undefined },
  showFilter: { value: undefined },
}

export const houseFilterValueLens = R.lensPath(['houseFilter', 'value'])
export const eventFilterValueLens = R.lensPath(['eventFilter', 'value'])
export const showFilterValueLens = R.lensPath(['showFilter', 'value'])

const setInitialState = user => R.set(houseFilterValueLens, R.path(['house', 'id'], user), initialState)

export const FilterContext = React.createContext(initialState)

export const FilterContextProvider = ({ user, children }) => {
  const [filters, setFilters] = useState(setInitialState(user))

  const setHouseValue = useCallback(id => setFilters(R.set(houseFilterValueLens, id)), [])
  const setEventValue = useCallback(id => setFilters(R.set(eventFilterValueLens, id)), [])
  const setShowValue = useCallback(v => setFilters(R.set(showFilterValueLens, v)), [])

  const ctx = {
    filters,
    setHouseValue,
    setEventValue,
    setShowValue,
  }

  return <FilterContext.Provider value={ctx}>{children}</FilterContext.Provider>
}
