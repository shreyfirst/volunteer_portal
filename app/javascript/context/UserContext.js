import React, { useState, useCallback } from 'react'

import * as R from 'ramda'

const preferenceLens = R.lensProp('preference')
const houseLens = R.lensProp('house')

export const UserContext = React.createContext(null)

export const UserContextProvider = ({ user, children }) => {
  const [currentUser, setCurrentUser] = useState(user)

  const setHouse = useCallback(house => setCurrentUser(R.set(houseLens, house)), [])
  const setPreference = useCallback(pref => setCurrentUser(R.set(preferenceLens, pref)), [])
  const clear = useCallback(() => setCurrentUser(undefined), [])

  const ctx = {
    currentUser,
    setHouse,
    setPreference,
    clear,
  }

  return <UserContext.Provider value={ctx}>{children}</UserContext.Provider>
}
