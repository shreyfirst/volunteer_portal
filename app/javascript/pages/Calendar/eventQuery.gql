#import "fragments/UserEntry.gql"
#import "fragments/EventEntry.gql"
#import "fragments/HouseEntry.gql"
#import "fragments/EventTypeEntry.gql"
#import "fragments/OrganizationEntry.gql"

query EventQuery($id: ID!) {
  currentUser {
    ...UserEntry
    isAdmin
  }
  event(id: $id) {
    ...EventEntry
    eventType {
      ...EventTypeEntry
    }
    users {
      ...UserEntry
    }
    house {
      ...HouseEntry
    }
    organization {
      ...OrganizationEntry
    }
  }
}
