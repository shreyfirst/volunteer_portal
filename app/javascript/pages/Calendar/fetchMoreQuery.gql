#import "fragments/EventEntry.gql"
#import "fragments/HouseEntry.gql"
#import "fragments/UserEntry.gql"

query FetchMoreQuery($after: Int, $before: Int) {
  events(after: $after, before: $before) {
    ...EventEntry
    signupCount
    house {
      ...HouseEntry
    }
    users {
      ...UserEntry
    }
  }
}
