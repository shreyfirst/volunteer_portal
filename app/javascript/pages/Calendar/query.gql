#import "fragments/UserEntry.gql"
#import "fragments/EventEntry.gql"
#import "fragments/HouseEntry.gql"
#import "fragments/UserEntry.gql"

query EventsQuery($after: Int, $before: Int, $houseId: ID) {
  currentUser {
    ...UserEntry
    isAdmin
    house {
      ...HouseEntry
    }
  }
  events(after: $after, before: $before, houseId: $houseId) {
    ...EventEntry
    signupCount
    house {
      ...HouseEntry
    }
    users {
      ...UserEntry
    }
  }
  houses(sortBy: NAME_ASC) {
    ...HouseEntry
  }
}
