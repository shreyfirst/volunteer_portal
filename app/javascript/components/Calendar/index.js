import React, { useContext } from 'react'
import moment from 'moment'
import * as R from 'ramda'
import BigCalendar from 'react-big-calendar'

import Event from 'components/Event'
import Layout from 'components/Layout'
import Loading from 'components/LoadingIcon'
import Toolbar from 'components/Toolbar'
import { FilterContext, UserContext } from '/context'

import 'style-loader!css-loader!react-big-calendar/lib/css/react-big-calendar.css'
import i18next from 'i18next'
import { withTranslation } from 'react-i18next'

const localizer = BigCalendar.momentLocalizer(moment)
BigCalendar.setLocalizer(localizer)

// BigCalendar needs inline styles :(
const styles = {
  cell: {
    backgroundColor: 'transparent',
    color: '#555',
  },
}

// Custom components given to BigCalendar
const calendarComponents = {
  toolbar: props => <Toolbar {...props} />,
  event: Event, // used by each view (Month, Day, Week)
}

const eventPropGetter = (_event, _start, _end, _isSelected) => ({
  style: styles.cell,
})

// Filter pipeline for selecting events to display on the calendar
const normalizeEvents = R.map(event =>
  R.merge(event, {
    start: new Date(event.startsAt),
    end: new Date(event.endsAt),
  })
)

const applyShowFilter = dataAndFilters => {
  const {
    event,
    currentUser,
    filters: { showFilter },
    isValid,
  } = dataAndFilters

  if (isValid && showFilter.value == 'mine') {
    return R.merge(dataAndFilters, { isValid: R.any(user => user.id === currentUser.id, event.users) })
  }

  return dataAndFilters
}

const applyEventFilter = dataAndFilters => {
  const {
    event,
    currentUser,
    filters: { eventFilter },
    isValid,
  } = dataAndFilters

  if (isValid) {
    switch (eventFilter.value) {
      case 'open':
        return R.merge(dataAndFilters, { isValid: event.users.length < event.capacity })
      case 'full':
        return R.merge(dataAndFilters, { isValid: event.users.length >= event.capacity })
      default:
        dataAndFilters
    }
  }

  return dataAndFilters
}

const applyHouseFilter = dataAndFilters => {
  const {
    event,
    filters: { houseFilter },
    isValid,
  } = dataAndFilters
  const showAll = houseFilter.value === 'all'

  if (isValid && !showAll) {
    return R.merge(dataAndFilters, { isValid: event.house.id == houseFilter.value })
  }

  return dataAndFilters
}

const filterPipeline = (currentUser, filters, isValid) =>
  R.pipe(
    event => ({ event, currentUser, filters, isValid }),
    applyShowFilter,
    applyEventFilter,
    applyHouseFilter,
    R.prop('isValid')
  )

const selectEvents = (events, currentUser, filters) =>
  R.pipe(
    R.filter(filterPipeline(currentUser, filters, true)),
    R.values,
    normalizeEvents
  )(events)

const Calendar = ({ loading, currentPath, events, togglePopover, loadMoreEvents }) => {
  const { currentUser } = useContext(UserContext)
  const { filters } = useContext(FilterContext)

  if (loading) return <Loading />

  return (
    <Layout currentPath={currentPath}>
      <BigCalendar
        events={selectEvents(events, currentUser, filters)}
        eventPropGetter={eventPropGetter}
        views={['month']}
        culture={i18next.language}
        components={calendarComponents}
        onSelectEvent={(event, e) => togglePopover('event', event, e.currentTarget)}
        onNavigate={loadMoreEvents}
        popup
        style={{ paddingBottom: 20 }}
      />
    </Layout>
  )
}

// Wrap `withTranslation` to listen to language change event
export default withTranslation()(Calendar)
