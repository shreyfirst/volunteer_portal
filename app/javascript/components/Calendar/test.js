import React from 'react'
import Calendar from './'
import { UserContextProvider, FilterContextProvider } from '/context'
import { shallow } from 'enzyme'

const currentUser = {
  id: '1',
  name: 'Craig Day',
  email: 'volunteer@example.com',
  photo: 'https://example.com/photo.jpg',
  isAdmin: true,
  house: {
    id: '1',
    identifier: 'san_francisco',
    name: 'San Francisco',
  },
}

const events = []
const eventPopover = {}
const togglePopover = () => {}
const loadMoreEvents = () => {}

test('loads', () => {
  const component = shallow(
    <UserContextProvider user={currentUser}>
      <FilterContextProvider user={currentUser}>
        <Calendar
          loading={false}
          currentPath={'/'}
          events={events}
          eventPopover={eventPopover}
          togglePopover={togglePopover}
          loadMoreEvents={loadMoreEvents}
        />
      </FilterContextProvider>
    </UserContextProvider>
  )

  expect(component).toMatchSnapshot()
})
