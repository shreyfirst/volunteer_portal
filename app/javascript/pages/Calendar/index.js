import React, { useContext } from 'react'
import { useQuery } from '@apollo/react-hooks'
import { compose, graphql } from 'react-apollo'
import { NetworkStatus } from 'apollo-client'
import { connect } from 'react-redux'
import * as R from 'ramda'
import moment from 'moment'

import { togglePopover, changeShowFilter, changeEventFilter, changeHouseFilter, calendarDateChange } from 'actions'

import Calendar from 'components/Calendar'
import EventPopover from 'components/EventPopover'

import EventsQuery from './query.gql'
import EventQuery from './eventQuery.gql'
import FetchMoreQuery from './fetchMoreQuery.gql'
import CreateSignupMutation from 'mutations/CreateSignupMutation.gql'
import DestroySignupMutation from 'mutations/DestroySignupMutation.gql'
import { UserContext, FilterContext, houseFilterValueLens } from '/context'

const fetchMoreEvents = (fetchMore, calendarDateChange, currentDate, newDate, scope) => {
  const startOfNewScope = moment(newDate).startOf(scope)

  const startOfNextScope =
    newDate > currentDate ? startOfNewScope.add(1, `${scope}s`) : startOfNewScope.subtract(1, `${scope}s`)

  const after = Number(startOfNewScope.format('X'))
  const before = Number(startOfNewScope.endOf(scope).format('X'))

  fetchMore({
    query: FetchMoreQuery,
    variables: { after, before },
    updateQuery: (previousResult, { fetchMoreResult }) => {
      if (!fetchMoreResult) {
        return previousResult
      }

      const existing = new Set(previousResult.events)

      R.forEach(e => existing.add(e), fetchMoreResult.events)

      return R.merge(previousResult, { events: [...existing] })
    },
  })

  calendarDateChange(newDate)

  return newDate
}

const EventPopoverForData = props => {
  const {
    popoverState,
    currentUser,
    onPopoverClose,
    createSignup,
    destroySignup,
    data: { networkStatus, event },
  } = props

  return (
    <EventPopover
      loading={networkStatus === NetworkStatus.loading}
      open
      anchorEl={popoverState.anchorEl}
      currentUser={currentUser}
      event={event}
      onPopoverClose={onPopoverClose}
      createSignupHandler={() => createSignup(event, currentUser)}
      destroySignupHandler={() => destroySignup(event, currentUser)}
    />
  )
}

const connectEventPopoverToData = graphql(EventQuery, {
  options: ({
    popoverState: {
      data: { id },
    },
  }) => ({
    variables: {
      id: id,
    },
  }),
})

const EventPopoverWithData = connectEventPopoverToData(EventPopoverForData)

const CalendarPage = ({
  locationBeforeTransitions,
  eventPopover,
  togglePopover,
  createSignup,
  destroySignup,
  calendarDateChange,
  calendarDate,
}) => {
  const { currentUser } = useContext(UserContext)
  const { filters } = useContext(FilterContext)
  const { data, loading, fetchMore } = useQuery(EventsQuery, {
    variables: {
      after: momentAfter,
      before: momentBefore,
      houseId: R.view(houseFilterValueLens, filters),
    },
  })

  const events = R.propOr([], 'events')(data)
  const houses = R.propOr([], 'houses')(data)

  return (
    <div>
      <Calendar
        loading={loading}
        currentPath={locationBeforeTransitions.pathname}
        events={events}
        houses={houses}
        currentUser={currentUser}
        eventPopover={eventPopover}
        togglePopover={togglePopover}
        createSignup={createSignup}
        destroySignup={destroySignup}
        loadMoreEvents={R.partial(fetchMoreEvents, [fetchMore, calendarDateChange, calendarDate])}
      />
      {eventPopover && (
        <EventPopoverWithData
          currentUser={currentUser}
          popoverState={eventPopover}
          onPopoverClose={togglePopover}
          createSignup={createSignup}
          destroySignup={destroySignup}
        />
      )}
    </div>
  )
}

const mapStateToProps = state => {
  const { popover, calendarDate } = state.model
  const { locationBeforeTransitions } = state.routing

  return {
    eventPopover: popover && popover.type === 'event' ? popover : null,
    calendarDate,
    locationBeforeTransitions,
  }
}

const momentAfter = Number(
  moment()
    .subtract(1, 'months')
    .startOf('month')
    .format('X')
)

const momentBefore = Number(
  moment()
    .add(1, 'months')
    .endOf('month')
    .format('X')
)

const updateEventsCache = (cache, eventChange, houseId) => {
  const variables = {
    after: momentAfter,
    before: momentBefore,
    houseId: houseId || 'current',
  }

  const prevCache = cache.readQuery({ query: EventsQuery, variables })

  const update = {
    ...prevCache,
    events: [
      ...prevCache.events.filter(event => event.id != eventChange.id),
      {
        ...prevCache.events.find(event => event.id == eventChange.id),
        users: eventChange.users,
        signupCount: eventChange.users.length,
      },
    ],
  }
  cache.writeQuery({ query: EventsQuery, variables, data: update })
}

const withData = compose(
  graphql(CreateSignupMutation, {
    props: ({ ownProps, mutate }) => ({
      createSignup: (event, currentUser) =>
        mutate({
          variables: { eventId: event.id },
          optimisticResponse: {
            __typename: 'Mutation',
            createSignup: {
              __typename: 'Signup',
              event: R.merge(event, {
                users: R.append(currentUser, event.users),
              }),
            },
          },
          update: (
            cache,
            {
              data: {
                createSignup: { event: eventChange },
              },
            }
          ) => {
            updateEventsCache(cache, eventChange, ownProps.filters.houseFilter.value)
          },
        }),
    }),
  }),
  graphql(DestroySignupMutation, {
    props: ({ ownProps, mutate }) => ({
      destroySignup: (event, currentUser) =>
        mutate({
          variables: { eventId: event.id, userId: currentUser.id },
          optimisticResponse: {
            __typename: 'Mutation',
            destroySignup: {
              __typename: 'Signup',
              event: R.merge(event, {
                users: R.reject(u => u.id === currentUser.id, event.users),
              }),
            },
          },
          update: (
            cache,
            {
              data: {
                destroySignup: { event: eventChange },
              },
            }
          ) => {
            updateEventsCache(cache, eventChange, ownProps.filters.houseFilter.value)
          },
        }),
    }),
  })
)

const withActions = connect(
  mapStateToProps,
  {
    togglePopover,
    calendarDateChange,
  }
)

export default withActions(withData(CalendarPage))
