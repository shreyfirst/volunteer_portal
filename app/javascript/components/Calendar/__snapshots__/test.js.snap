// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads 1`] = `
<ContextProvider
  value={
    Object {
      "clear": [Function],
      "currentUser": Object {
        "email": "volunteer@example.com",
        "id": "1",
        "isAdmin": true,
        "name": "Craig Day",
        "house": Object {
          "id": "1",
          "identifier": "san_francisco",
          "name": "San Francisco",
        },
        "photo": "https://example.com/photo.jpg",
      },
      "setHouse": [Function],
      "setPreference": [Function],
    }
  }
>
  <FilterContextProvider
    user={
      Object {
        "email": "volunteer@example.com",
        "id": "1",
        "isAdmin": true,
        "name": "Craig Day",
        "house": Object {
          "id": "1",
          "identifier": "san_francisco",
          "name": "San Francisco",
        },
        "photo": "https://example.com/photo.jpg",
      }
    }
  >
    <withI18nextTranslation(Calendar)
      currentPath="/"
      eventPopover={Object {}}
      events={Array []}
      loadMoreEvents={[Function]}
      loading={false}
      togglePopover={[Function]}
    />
  </FilterContextProvider>
</ContextProvider>
`;
